.github/workflows/tests.yml

```diff
@@ -119,24 +131,91 @@ jobs:
 
       - name: Start Covalent dispatcher server
         run: covalent start
+               
+      - name: Run SDK tests and measure coverage
+        if: ${{ steps.working-directory.outputs.sdk}} == 'true'
+        uses: nick-fields/retry@v2
+        with:
+          timeout_minutes: 15
+          max_attempts: 2
+          on_retry_command: rm -f ~/.config/covalent/covalent.conf
+          command: |
+            PYTHONPATH=$PWD/ pytest -vv tests/covalent_tests  --cov=covalent --cov-config=.coveragerc
+            
+      - name: Generate SDK coverage report
+        run: coverage xml -o sdk_coverage.xml
+
+      - name: Run dispatcher tests and measure coverage
+        if: ${{ steps.working-directory.outputs.dispatcher }} == 'true'
+        uses: nick-fields/retry@v2
+        with:
+          timeout_minutes: 15
+          max_attempts: 2
+          on_retry_command: rm -f ~/.config/covalent/covalent.conf
+          command: |
+            PYTHONPATH=$PWD/ pytest -vv tests/covalent_dispatcher_tests  --cov=covalent_dispatcher --cov-config=.coveragerc
+      
+      - name: Generate dispatcher coverage report
+        run: coverage xml -o dispatcher_coverage.xml
+
+      - name: Run functional tests and measure coverage
+        if: ${{ steps.working-directory.outputs.functional_tests }} == 'true'
+        uses: nick-fields/retry@v2
+        with:
+          timeout_minutes: 15
+          max_attempts: 2
+          on_retry_command: rm -f ~/.config/covalent/covalent.conf
+          command: |
+            PYTHONPATH=$PWD/ pytest -vv tests/functional_tests  --cov=covalent --cov=covalent_dispatcher --cov-config=.coveragerc
+      
+      - name: Generate functional test coverage report
+        run: coverage xml -o functional_tests_coverage.xml
 
-      - name: Run tests and measure coverage
+      - name: Run UI backend tests and measure coverage
+        if: ${{ steps.working-directory.outputs.covalent_ui }} == 'true' && false  && [ -d "tests/covalent_ui" ]
```

Comment:
I really like your idea here of checking if the test folder exists, it's much more elegant than my suggestion of just putting in a `false`. 2 points:
1. I think the path should be `tests/covalent_ui_tests`
2. Can you double check that the POSIX test syntax can be used like this in the workflow `if` field? I don't think that will work, but maybe. [This community post](https://github.community/t/run-step-if-file-exists/16445) has some good suggestions for ways to test if a path exists.

Reply:
Thank you! I'll check out the post for the suggestions on the right syntax to use.

Reply:
@scottwn I chained an 'or' statement to the pytest command in #L182 of this commit https://github.com/AgnostiqHQ/covalent/pull/822/commits/3be0f4019ff225bbeda9e6103d57e5956f7fb996 to skip execution of the step if the path does not exist and/or if the command fails.

Reply:
Looks great. Can you make these coverage & upload steps depend on the output of `Run UI backend tests and measure coverage` so that we can remove those `if: false` as well?

https://github.com/AgnostiqHQ/covalent/pull/822/files#diff-1db27d93186e46d3b441ece35801b244db8ee144ff1405ca27a163bfe878957fR185
https://github.com/AgnostiqHQ/covalent/pull/822/files#diff-1db27d93186e46d3b441ece35801b244db8ee144ff1405ca27a163bfe878957fR213

Reply:
I have replaced the `if: false` condition with a chain of statements to check for the working-directory output, test-output, and exit code in the `Run UI backend tests and measure coverage` step. 
https://github.com/AgnostiqHQ/covalent/blob/6c84c6de64513d76d0a54a0f3bc819f530c5b001/.github/workflows/tests.yml#L189
https://github.com/AgnostiqHQ/covalent/blob/6c84c6de64513d76d0a54a0f3bc819f530c5b001/.github/workflows/tests.yml#L219